File: src/StickedWords.Domain/Models/Exercises/GuessResult.cs
namespace StickedWords.Domain.Models.Exercises;

public record GuessResult : IExerciseResult
{
    private GuessResult() { }

    public Verdict Result { get; init; }

    public string? CorrectTranslation { get; init; }

    public required bool IsExpired { get; init; }

    public static GuessResult Correct(LearningSessionState sessionState)
    {
        return new()
        {
            Result = Verdict.Correct,
            IsExpired = sessionState is LearningSessionState.Expired
        };
    }

    public static GuessResult Wrong(FlashCardWord translation, LearningSessionState sessionState)
    {
        return new()
        {
            Result = Verdict.Wrong,
            CorrectTranslation = translation.Word,
            IsExpired = sessionState is LearningSessionState.Expired
        };
    }
}

File: src/StickedWords.UI.React/src/models/exercises/GuessResult.ts
export enum Verdict {
  None = "None",
  Correct = "Correct",
  Wrong = "Wrong"
}

export class GuessResult {
  
  result!: Verdict;
  correctTranslation?: string;
  isExpired!: boolean;

  static fromJson(json: any): GuessResult {
    const result = new GuessResult();

    result.result = GuessResult.mapVerdict(json.result);
    result.correctTranslation = json.correctTranslation ?? null;
    result.isExpired = json.isExpired ?? false;

    return result;
  }

  private static mapVerdict(src: string): Verdict {
    if (!src) {
      return Verdict.None;
    }

    return Object.values(Verdict).includes(src as Verdict)
      ? src as Verdict
      : Verdict.None;
  }
}
